### PR TITLE
fix(*): interface ResourceSpec depends on proto.Message

### PR DESCRIPTION
**Pull request by lobkovilya**
We want to stop using protobuf to define new policies. This PR removes dependency on `proto.Message` from `core_model.ResourceSpec` interface. It also replace all proto functions (like Equal, IsEmpty) with universal functions from `core_model`.

Signed-off-by: Ilya Lobkov <ilya.lobkov@konghq.com>

### Checklist prior to review

- [X] Link to docs PR or issue --
- [X] Link to UI issue or PR --
- [X] Is the [issue worked on linked][1]? --
- [X] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [X] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Unit Tests --
- [ ] E2E Tests --
- [ ] Manual Universal Tests --
- [ ] Manual Kubernetes Tests --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? -- No
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? -- No
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by lahabana** [comment]
Please take bigger caution in your git commit body.

For example here:
- there's no body in the commit here 
- maybe instead for `fix(*)` it should be `fix(policy)`. Have you rolled together all these PR in a common changelog item with `> Changelog` ?
- In the PR description you explain what you do. What is usually most important in commits is to explain *why* you are doing something.

File: pkg/core/resources/model/utils.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	fn      func(T, any) error
+}
+
+func (m *unmarshal[T]) ResourceSpec(src T, spec ResourceSpec) error {
+	if msg, ok := spec.(proto.Message); ok {
+		return m.protoFn(src, msg)
+	} else {
+		return m.fn(src, spec)
+	}
+}
```
Comment (lahabana) [inline pkg/core/resources/model/utils.go:85]: I'm not sure generics are useful here if you end up with an if/else on the type.

**Comment by slonka** [comment]
> maybe instead for fix(*) it should be fix(policy)

Shouldn't this indicate the component? It's one of `kumacp`, `kumadp`, `kumactl` etc. Also thanks for reminding us about the commit message and `> Changelog`

**Review by lahabana — CHANGES_REQUESTED** [review]
I'm 99% sure generics here are just bringing more complexity. We can chat about it live if you want

**Comment by lobkovilya** [comment]
> there's no body in the commit here

If we want a body in commits this check should be enforced, but I don't understand why we need this if in the end, we stash all commits before the merge.

> maybe instead for fix(*) it should be fix(policy)

As @slonka said, it's about components, not about feature

> Have you rolled together all these PR in a common changelog item with > Changelog

Yes, I'm not done with migration from proto, I'll add 1 changelog item for all relative PRs

File: pkg/core/resources/model/utils.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/core/resources/model/utils.go:67]: Removed

**Comment by kleinfreund** [comment]
> If we want a body in commits this check should be enforced, but I don't understand why we need this if in the end, we stash all commits before the merge.

@lobkovilya With a good commit hygiene squashing commits won’t affect the meaning of the final set of commits. The merge commit won’t have much content but the commits from your branch will still end up in the commit history of the default branch with their commit bodies intact. It’s pretty nice to follow changes with descriptive commit bodies.

I also find it much easier to detail a change set in a commit body and then just copy this into the PR description (and possibly expand on it).

File: pkg/core/resources/model/utils.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
+package model
+
+import (
+	"encoding/json"
+	"path"
+	"reflect"
+
+	"github.com/ghodss/yaml"
```
Comment (lahabana) [inline pkg/core/resources/model/utils.go:8]: Isn't yaml2 the right yaml to use (I always get confused :) )

**Review by lahabana — COMMENTED** [review]
Intuitively looks good but I'm short on time to do a proper review

**Review by lahabana — APPROVED** [review]

**Comment by lahabana** [comment]
@lobkovilya what are we doing with this? Should it be merged?

File: pkg/core/resources/model/utils.go (hunk 2 above)
Comment (lobkovilya) [reply pkg/core/resources/model/utils.go:8]: We have 60 matches for `"github.com/ghodss/yaml"` and 5 matches for `"gopkg.in/yaml.v2"`, I don't know if there are any limitations why we can't stick to 1 library. I can create an issue for that

Comment (lobkovilya) [reply pkg/core/resources/model/utils.go:8]: https://github.com/kumahq/kuma/issues/5192

**Comment by lahabana** [comment]
@lobkovilya  merging then? :)

**Pull request merged**